cmd/kubeadm/app/cmd/reset.go

```diff
@@ -55,99 +56,109 @@ type Reset struct{}
 
 func NewReset(skipPreFlight bool) (*Reset, error) {
 	if !skipPreFlight {
-		fmt.Println("Running pre-flight checks")
-		err := preflight.RunResetCheck()
-		if err != nil {
+		fmt.Println("[preflight] Running pre-flight checks...")
+
+		if err := preflight.RunResetCheck(); err != nil {
 			return nil, &preflight.PreFlightError{Msg: err.Error()}
 		}
 	} else {
-		fmt.Println("Skipping pre-flight checks")
+		fmt.Println("[preflight] Skipping pre-flight checks...")
 	}
 
 	return &Reset{}, nil
 }
 
-// cleanDir removes everything in a directory, but not the directory itself:
-func cleanDir(path string) {
+// Run reverts any changes made to this host by "kubeadm init" or "kubeadm join".
+func (r *Reset) Run(out io.Writer) error {
+	serviceToStop := "kubelet"
+	initSystem, err := initsystem.GetInitSystem()
+	if err != nil {
+		fmt.Printf("[reset] Failed to detect init system and stop the kubelet service: %v\n", err)
+	} else {
+		fmt.Printf("[reset] Stopping the %s service...\n", serviceToStop)
+		if err := initSystem.ServiceStop(serviceToStop); err != nil {
+			fmt.Printf("[reset] Failed to stop the %s service", serviceToStop)
+		}
+	}
+
+	fmt.Println("[reset] Unmounting directories in /var/lib/kubelet...")
+	umountDirsCmd := "cat /proc/mounts | awk '{print $2}' | grep '/var/lib/kubelet' | xargs -r umount"
+	umountOutputBytes, err := exec.Command("sh", "-c", umountDirsCmd).Output()
+	if err != nil {
+		fmt.Printf("[reset] Failed to unmount directories in /var/lib/kubelet: %s\n", string(umountOutputBytes))
+	}
+
+	resetConfigDir()
+
+	dirsToClean := []string{"/var/lib/kubelet", "/var/lib/etcd", "/etc/cni/net.d"}
+	fmt.Printf("[reset] Deleting contents of stateful directories: %v\n", dirsToClean)
+	for _, dir := range dirsToClean {
+		cleanDir(dir)
+	}
+
+	dockerCheck := preflight.ServiceCheck{Service: "docker"}
+	if warnings, errors := dockerCheck.Check(); len(warnings) == 0 && len(errors) == 0 {
+		fmt.Println("[reset] Stopping all running docker containers...")
+		if err := exec.Command("sh", "-c", "docker ps | grep 'k8s_' | awk '{print $1}' | xargs -r docker rm --force --volumes").Run(); err != nil {
```

Comment:
we don't support other runtimes than docker for the moment, so that's not a problem for now.
this code just moved from below